lib/web_ui/lib/src/engine/initialization.dart

```diff
@@ -144,8 +215,93 @@ void initializeEngine() {
     }
   };
 
+  // This needs to be after `initializeEngine` because that is where the
+  // canvaskit script is added to the page.
+  if (useCanvasKit) {
+    await initializeCanvasKit();
+  }
+
+  assetManager ??= const AssetManager();
+  await _setAssetManager(assetManager);
+  if (useCanvasKit) {
+    await skiaFontCollection.ensureFontsLoaded();
+  } else {
+    await _fontCollection!.ensureFontsLoaded();
+  }
+  _initializationState = DebugEngineInitializationState.initializedServices;
+}
+
+/// Initializes the UI surfaces for the Flutter framework to render to.
+///
+/// Must be called after [initializeEngineServices].
+///
+/// This function will start altering the HTML structure of the page. If used
+/// in an add-to-app scenario, the host page is expected to prepare for Flutter
+/// UI appearing on screen prior to calling this function.
+Future<void> initializeEngineUi() async {
+  if (_initializationState != DebugEngineInitializationState.initializedServices) {
+    assert(() {
+      throw StateError(
+        'Invalid engine initialization state. `initializeEngineUi` was '
+        'called while the engine initialization state was '
+        '"$_initializationState". `initializeEngineUi` can only be called '
+        'when the engine is in state '
+        '"${DebugEngineInitializationState.initializedServices}".'
+      );
+    }());
+    return;
+  }
+  _initializationState = DebugEngineInitializationState.initializingUi;
+
   Keyboard.initialize(onMacOs: operatingSystem == OperatingSystem.macOs);
   MouseCursor.initialize();
+  ensureFlutterViewEmbedderInitialized();
+
+  if (useCanvasKit) {
+    /// Add a Skia scene host.
+    skiaSceneHost = html.Element.tag('flt-scene');
+    flutterViewEmbedder.renderScene(skiaSceneHost);
+  }
+  _initializationState = DebugEngineInitializationState.initialized;
+}
+
+AssetManager get assetManager => _assetManager!;
+AssetManager? _assetManager;
+
+FontCollection get fontCollection => _fontCollection!;
+FontCollection? _fontCollection;
+
+Future<void> _setAssetManager(AssetManager assetManager) async {
+  // ignore: unnecessary_null_comparison
+  assert(assetManager != null, 'Cannot set assetManager to null');
+  if (assetManager == _assetManager) {
+    return;
+  }
+
+  _assetManager = assetManager;
+
+  if (useCanvasKit) {
+    ensureSkiaFontCollectionInitialized();
+  } else {
+    _fontCollection ??= FontCollection();
+    _fontCollection!.clear();
+  }
+
+  if (_assetManager != null) {
+    if (useCanvasKit) {
+      await skiaFontCollection.registerFonts(_assetManager!);
+    } else {
+      await _fontCollection!.registerFonts(_assetManager!);
+    }
```

Comment:
Duplicate code (with line 226)?

Reply:
No, the code above is only waiting for the fonts to load, it doesn't register fonts.

Reply:
😱 my eyes tricked me 😅